File: sdk/python/examples/jira_onboarding.py
import json
import os
from typing import Any, Dict

from cerebro_sdk import Client, IntegrationClient


def build_issue_claims(integration: IntegrationClient, issue: Dict[str, Any]) -> list[Dict[str, Any]]:
    issue_key = str(issue["key"]).strip()
    issue_ref = integration.ref("ticket", issue_key, issue_key)
    claims = [
        integration.exists(
            issue_ref,
            source_event_id=str(issue.get("event_id", "")).strip() or None,
        ),
        integration.attr(issue_ref, "summary", str(issue["summary"]).strip()),
        integration.attr(issue_ref, "status", str(issue["status"]).strip()),
    ]

    project_key = str(issue.get("project_key", "")).strip()
    if project_key:
        claims.append(
            integration.rel(
                issue_ref,
                "belongs_to",
                integration.ref("project", project_key, project_key),
            )
        )

    assignee_email = str(issue.get("assignee_email", "")).strip()
    if assignee_email:
        claims.append(
            integration.rel(
                issue_ref,
                "assigned_to",
                integration.ref("user", assignee_email, assignee_email),
            )
        )

    reporter_email = str(issue.get("reporter_email", "")).strip()
    if reporter_email:
        claims.append(
            integration.rel(
                issue_ref,
                "reported_by",
                integration.ref("user", reporter_email, reporter_email),
            )
        )

    priority = str(issue.get("priority", "")).strip()
    if priority:
        claims.append(integration.attr(issue_ref, "priority", priority))

    return claims


def onboard_issue(base_url: str, api_key: str, tenant_id: str, runtime_id: str, issue: Dict[str, Any]) -> Dict[str, Any]:
    client = Client(base_url=base_url, api_key=api_key or None)
    integration = client.integration(runtime_id=runtime_id, tenant_id=tenant_id, integration="jira")
    runtime_config = {}
    workspace = str(issue.get("workspace", "")).strip()
    if workspace:
        runtime_config["workspace"] = workspace
    integration.ensure_runtime(runtime_config)
    claims = build_issue_claims(integration, issue)
    write_result = integration.write_claims(claims)
    claim_result = integration.list_claims(
        {
            "subject_urn": claims[0]["subject_urn"],
            "limit": 20,
        }
    )
    return {
        "write_result": write_result,
        "claims": claim_result.get("claims", []),
    }


def main() -> None:
    base_url = os.environ.get("CEREBRO_BASE_URL", "").strip()
    if not base_url:
        raise RuntimeError("CEREBRO_BASE_URL is required")

    result = onboard_issue(
        base_url=base_url,
        api_key=os.environ.get("CEREBRO_API_KEY", "").strip(),
        tenant_id=os.environ.get("CEREBRO_TENANT_ID", "writer").strip(),
        runtime_id=os.environ.get("CEREBRO_RUNTIME_ID", "writer-jira").strip(),
        issue={
            "workspace": os.environ.get("JIRA_WORKSPACE", "writer").strip(),
            "event_id": os.environ.get("JIRA_EVENT_ID", "jira-event-1").strip(),
            "project_key": os.environ.get("JIRA_PROJECT_KEY", "ENG").strip(),
            "key": os.environ.get("JIRA_ISSUE_KEY", "ENG-123").strip(),
            "summary": os.environ.get("JIRA_ISSUE_SUMMARY", "Claim-first Jira onboarding example").strip(),
            "status": os.environ.get("JIRA_ISSUE_STATUS", "in_progress").strip(),
            "priority": os.environ.get("JIRA_ISSUE_PRIORITY", "high").strip(),
            "assignee_email": os.environ.get("JIRA_ASSIGNEE_EMAIL", "alice@writer.com").strip(),
            "reporter_email": os.environ.get("JIRA_REPORTER_EMAIL", "bob@writer.com").strip(),
        },
    )
    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: sdk/typescript/examples/jira_onboarding.ts
import { Client, type Claim, type IntegrationClient } from "../src/index.js";

declare const process: {
  env: Record<string, string | undefined>;
  argv?: string[];
};

export interface JiraIssuePayload {
  workspace?: string;
  eventId?: string;
  projectKey?: string;
  key?: string | null;
  summary?: string | null;
  status?: string | null;
  priority?: string;
  assigneeEmail?: string;
  reporterEmail?: string;
}

export interface OnboardIssueOptions {
  baseUrl: string;
  apiKey?: string;
  tenantId: string;
  runtimeId: string;
  issue: JiraIssuePayload;
}

export function buildIssueClaims(integration: IntegrationClient, issue: JiraIssuePayload): Claim[] {
  const issueKey = requireValue(issue.key, "issue.key");
  const summary = requireValue(issue.summary, "issue.summary");
  const status = requireValue(issue.status, "issue.status");

  const issueRef = integration.ref("ticket", issueKey, issueKey);
  const sourceEventId = issue.eventId?.trim();
  const sharedOptions = sourceEventId ? { source_event_id: sourceEventId } : {};
  const claims: Claim[] = [
    integration.exists(issueRef, sharedOptions),
    integration.attr(issueRef, "summary", summary, sharedOptions),
    integration.attr(issueRef, "status", status, sharedOptions),
  ];

  const projectKey = issue.projectKey?.trim();
  if (projectKey) {
    claims.push(
      integration.rel(
        issueRef,
        "belongs_to",
        integration.ref("project", projectKey, projectKey),
        sharedOptions,
      ),
    );
  }

  const assigneeEmail = issue.assigneeEmail?.trim();
  if (assigneeEmail) {
    claims.push(
      integration.rel(
        issueRef,
        "assigned_to",
        integration.ref("user", assigneeEmail, assigneeEmail),
        sharedOptions,
      ),
    );
  }

  const reporterEmail = issue.reporterEmail?.trim();
  if (reporterEmail) {
    claims.push(
      integration.rel(
        issueRef,
        "reported_by",
        integration.ref("user", reporterEmail, reporterEmail),
        sharedOptions,
      ),
    );
  }

  const priority = issue.priority?.trim();
  if (priority) {
    claims.push(integration.attr(issueRef, "priority", priority, sharedOptions));
  }

  return claims;
}

export async function onboardIssue(options: OnboardIssueOptions): Promise<Record<string, unknown>> {
  const client = new Client({
    baseUrl: options.baseUrl,
    apiKey: options.apiKey?.trim() || undefined,
  });
  const integration = client.integration({
    runtimeId: options.runtimeId.trim(),
    tenantId: options.tenantId.trim(),
    integration: "jira",
  });
  const runtimeConfig: Record<string, string> = {};
  const workspace = options.issue.workspace?.trim();
  if (workspace) {
    runtimeConfig.workspace = workspace;
  }
  await integration.ensureRuntime(runtimeConfig);
  const claims = buildIssueClaims(integration, options.issue);
  const writeResult = await integration.writeClaims(claims);
  const subjectUrn = claims[0]?.subject_urn ?? "";
  const claimResult = await integration.listClaims(subjectUrn ? { subject_urn: subjectUrn, limit: 20 } : { limit: 20 });
  return {
    writeResult,
    claims: Array.isArray(claimResult["claims"]) ? claimResult["claims"] : [],
  };
}

async function main(): Promise<void> {
  const baseUrl = process.env.CEREBRO_BASE_URL?.trim() ?? "";
  if (!baseUrl) {
    throw new Error("CEREBRO_BASE_URL is required");
  }
  const result = await onboardIssue({
    baseUrl,
    apiKey: process.env.CEREBRO_API_KEY?.trim(),
    tenantId: process.env.CEREBRO_TENANT_ID?.trim() || "writer",
    runtimeId: process.env.CEREBRO_RUNTIME_ID?.trim() || "writer-jira",
    issue: {
      workspace: process.env.JIRA_WORKSPACE?.trim() || "writer",
      eventId: process.env.JIRA_EVENT_ID?.trim() || "jira-event-1",
      projectKey: process.env.JIRA_PROJECT_KEY?.trim() || "ENG",
      key: process.env.JIRA_ISSUE_KEY?.trim() || "ENG-123",
      summary: process.env.JIRA_ISSUE_SUMMARY?.trim() || "Claim-first Jira onboarding example",
      status: process.env.JIRA_ISSUE_STATUS?.trim() || "in_progress",
      priority: process.env.JIRA_ISSUE_PRIORITY?.trim() || "high",
      assigneeEmail: process.env.JIRA_ASSIGNEE_EMAIL?.trim() || "alice@writer.com",
      reporterEmail: process.env.JIRA_REPORTER_EMAIL?.trim() || "bob@writer.com",
    },
  });
  console.log(JSON.stringify(result, null, 2));
}

if (typeof process !== "undefined" && process.env) {
  const entrypoint = process.argv?.[1] ?? "";
  if (entrypoint.endsWith("jira_onboarding.ts") || entrypoint.endsWith("jira_onboarding.js")) {
    void main();
  }
}

function requireValue(value: string | undefined | null, name: string): string {
  if (typeof value !== "string") {
    throw new Error(`${name} is required`);
  }
  const normalized = value.trim();
  if (!normalized) {
    throw new Error(`${name} is required`);
  }
  return normalized;
}
